examples/autoignition-mpi.py

```diff
@@ -222,39 +222,70 @@ def my_rhs(t, state):
                                boundaries=boundaries, eos=eos)
                 + eos.get_species_source_terms(state))
 
-    def my_checkpoint(step, t, dt, state):
-        reaction_rates = eos.get_production_rates(state)
-        viz_fields = [("reaction_rates", reaction_rates)]
-        return sim_checkpoint(discr, visualizer, eos, cv=state,
-                              vizname=casename, step=step,
-                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
-                              constant_cfl=constant_cfl, comm=comm,
-                              viz_fields=viz_fields)
-
-    try:
-        (current_step, current_t, current_state) = \
-            advance_state(rhs=my_rhs, timestepper=timestepper,
-                checkpoint=my_checkpoint,
-                get_timestep=get_timestep, state=current_state,
-                t=current_t, t_final=t_final)
-    except ExactSolutionMismatch as ex:
-        error_state = True
-        current_step = ex.step
-        current_t = ex.t
-        current_state = ex.state
+    def my_checkpoint(step, t, dt, state, force=False):
+        from mirgecom.simutil import check_step
+        do_status = force or check_step(step=step, interval=nstatus)
+        do_viz = force or check_step(step=step, interval=nviz)
+        do_health = force or check_step(step=step, interval=nhealth)
```

Comment:
This feels like multiple functions jammed into one. Would it be better to allow registering multiple callbacks and making it such that each of them can, most often, be a library function?

Reply:
It totally is!  If the user could write an arbitrary number of callbacks and register them as pre or post step callbacks? That'd be super cool. The callbacks would need to be able to do i/o, collectives, and terminate the simulation.  If they could modify the state, that would be super-dee duper awesome, but not required currently.  Maybe filtering?